src/consumer_credit_policy.ts

```diff
@@ -0,0 +1,65 @@
+export type CreditRequestWrapper = (howMany: number) => Promise<void>
+
+export abstract class ConsumerCreditPolicy {
+  constructor(protected readonly onStartup: number) {}
+
+  public onChunkReceived(_requestWrapper: CreditRequestWrapper) {
+    return
+  }
+  public onChunkProgress(_idx: number, _total: number, _requestWrapper: CreditRequestWrapper) {
+    return
+  }
+  public onChunkCompleted(_requestWrapper: CreditRequestWrapper) {
+    return
+  }
+
+  public onSubscription() {
+    return this.onStartup
+  }
+}
+
+class NewCreditsOnChunkReceived extends ConsumerCreditPolicy {
+  constructor(onStartup: number = 1, private readonly onRenewal: number = 1) {
+    super(onStartup)
+  }
+
+  public async onChunkReceived(requestWrapper: CreditRequestWrapper) {
+    await requestWrapper(this.onRenewal)
+  }
+
+  public onSubscription(): number {
+    return this.onStartup
+  }
+}
+
+class NewCreditsOnChunkProgress extends ConsumerCreditPolicy {
+  constructor(onStartup: number = 1, private readonly ratio: number = 0.5, private readonly howMany: number = 1) {
```

Comment:
Maybe rename `howMany` to `onRenewal`? It would be more consistent and it's easier to understand what is happening, at least for me.

Reply:
onRenewal for me is a function/handler, (same as startup) usally in node onSoemthig indicate an handler, it's more `startFrom` and `step`